src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py

```diff
@@ -0,0 +1,200 @@
+    FormulaStep,
+    MetricFetcher,
+    Multiplier,
+    OpenParen,
+    Subtractor,
+)
+
+_operator_precedence = {
+    "(": 0,
+    "/": 1,
+    "*": 2,
+    "-": 3,
+    "+": 4,
+    ")": 5,
+}
+
+
+class FormulaEngine:
+    """A post-fix formula engine that operates on `Sample` receivers.
+
+    Operators and metrics need to be pushed into the engine in in-fix order, and they
+    get rearranged into post-fix in the engine.  This is done using the [Shunting yard
+    algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
+
+    Example:
+        To create an engine that adds the latest entries from two receivers, the
+        following calls need to be made:
+
+        ```python
+        engine = FormulaEngine()
+        engine.push_metric("metric_1", receiver_1)
+        engine.push_oper("+")
+        engine.push_metric("metric_2", receiver_2)
+        engine.finalize()
+        ```
+
+        and then every call to `engine.apply()` would fetch a value from each receiver,
+        add the values and return the result.
+    """
+
+    def __init__(
+        self,
+    ) -> None:
+        """Create a `FormulaEngine` instance."""
+        self._steps: List[FormulaStep] = []
+        self._build_stack: List[FormulaStep] = []
+        self._metric_fetchers: Dict[str, MetricFetcher] = {}
+        self._first_run = True
+
+    def push_oper(self, oper: str) -> None:
+        """Push an operator into the engine.
+
+        Args:
+            oper: One of these strings - "+", "-", "*", "/", "(", ")"
+        """
+        if self._build_stack and oper != "(":
+            op_prec = _operator_precedence[oper]
+            while self._build_stack:
+                prev_step = self._build_stack[-1]
+                if op_prec <= _operator_precedence[repr(prev_step)]:
+                    break
+                if oper == ")" and repr(prev_step) == "(":
+                    self._build_stack.pop()
+                    break
+                if repr(prev_step) == "(":
+                    break
+                self._steps.append(prev_step)
+                self._build_stack.pop()
+
+        if oper == "+":
+            self._build_stack.append(Adder())
+        elif oper == "-":
+            self._build_stack.append(Subtractor())
+        elif oper == "*":
+            self._build_stack.append(Multiplier())
+        elif oper == "/":
+            self._build_stack.append(Divider())
+        elif oper == "(":
+            self._build_stack.append(OpenParen())
+
+    def push_metric(self, name: str, data_stream: Receiver[Sample]) -> None:
+        """Push a metric receiver into the engine.
+
+        Args:
+            name: A name for the metric.
+            data_stream: A receiver to fetch this metric from.
+        """
+        if name not in self._metric_fetchers:
+            fetcher = MetricFetcher(name, data_stream)
+            self._metric_fetchers[name] = fetcher
+        else:
+            fetcher = self._metric_fetchers[name]
```

Comment:
`dict.setdefault()` is your friend :hugs: 

```suggestion
        fetcher = self._metric_fetchers.setdefault(name, MetricFetcher(name, data_stream))
```